Dev/wucke13 refactor 3

wucke13 opened the pull request:
### Pull Request Overview

This pull request adds recursive function calls

### Testing Strategy

This pull request was tested `cargo test`

### TODO or Help Wanted

This pull request still needs additional vetting

### Formatting

- [ ] Ran `cargo fmt`
- [ ] Ran `cargo check`
- [ ] Ran `cargo build`
- [ ] Ran `cargo doc`
- [ ] Ran `nix fmt`
- [ ] Ran `treefmt`

### Github Issue

This pull request closes #63, #39

george-cosma commented on src/validation/code.rs:
```diff
@@ -96,6 +107,35 @@ fn read_instructions(
+            }
+            // call [t1*] -> [t2*]
+            CALL => {
+                // TODO get function params, reverse iter over it, for each param pop that value from the value_stack
```
I think we can remove this TODO, right?

george-cosma commented on src/validation/code.rs:
```diff
@@ -96,6 +107,35 @@ fn read_instructions(
+                let ret_vals_start = value_stack.len() - this_func_ty.returns.valtypes.len();
+                let _remaining_locals = value_stack.drain(..ret_vals_start);
+
+                // TODO(george-cosma): There is a bug here. Because there is another op code after this reteurn (an
+                // `end`), I have to let this function consume it to work for the tests.
```
As discussed previously, this is likely not a bug, since there is no evidence two `end`s  will ever be emitted by a compiler. 

The issue comes when a hand-made `wat` file which contains two `end`s is passed to the interpreter. In this case, after finishing with validation I propose we check that we reached the end of the function (all bytes were consumed).

george-cosma commented on src/validation/code.rs:
```diff
@@ -96,6 +107,35 @@ fn read_instructions(
+                let this_func_ty = &fn_types[type_idx_of_fn[this_function_idx]];
+
+                if value_stack.len() < this_func_ty.returns.valtypes.len() {
+                    return Err(Error::EndInvalidValueStack); // TODO(george-cosma): Is this the right error?
```
Needs to be resolved

wucke13 replied:
LGTM!

george-cosma commented on src/execution/value_stack.rs:
```diff
@@ -12,81 +13,147 @@ use crate::unreachable_validated;
+    pub func_idx: FuncIdx,
+
+    /// Local varaiables such as parameters for this [CallFrame]'s function
+    /// TODO(george-cosma): Do we want "locals" to be a Box<\[Value\]>? Do we want two layers of indirection when getting locals?
+    pub locals: Locals,
```
This comments should be resolved

wucke13 replied:
I think we can leave it as is. Currently `Locals` is a one-field struct over a `Box<[Value]>`, so I would expect the compiler to not actually introduced a second level of indirection.

george-cosma replied:
I've noticed that we rely a lot on the compiler doing smart optimizations, yet we never check for it explicitly. Perhaps we should keep track of our assumptions, and eventually verify them (if we can?)

wucke13 replied:
That definitely is a good though. Though there are so many micro decisions that I would like refactor anyhow once we have the base set of features, that I'm not sure if it makes sense to spent the effort on tracking our assumptions.

One way of doing this could be writing unit tests that just assert the size of types, i.e. in this case the size of `Locals` to be that of an `usize`, or more precisely of a `Box<[ValTypes]>`.

george-cosma replied:
I agree with you that it is early. I suggest we start investigating this matter further after we have stabilized our architecture

george-cosma commented on src/execution/value_stack.rs:
```diff
@@ -12,81 +13,147 @@ use crate::unreachable_validated;
+    /// Pop a [CallFrame] from the call stack
+    // TODO figure out if this has to change the value stack
+
+    pub fn pop_stackframe(&mut self) -> usize {
+        // TODO maybe manipulate the value stack
+        let stack_vals = self.frames.last().unwrap_validated().stack_values;
+        let return_addr = self.frames.pop().unwrap_validated().return_addr;
+
+        // Move ownership of previous stackframe's values to the previous stackframe
+        // In theroy, this should be only the return values of the previous stack frame.
+        if let Some(callframe) = self.frames.last_mut() {
+            callframe.stack_values += stack_vals;
         }
+
+        return_addr
+    }
+
+    /// Pop a stackframe from the call stack
+    // TODO figure out if this has to change the value stack
+    pub fn push_stackframe(&mut self, func_idx: FuncIdx, locals: Locals, return_addr: usize) {
+        // TODO maybe manipulate the value stack
+        self.frames.push(CallFrame {
+            func_idx,
+            locals,
+            return_addr,
+            stack_values: 0,
+        })
+    }
```
I'm not sure I understand these TODOs

george-cosma commented on src/execution/mod.rs:
```diff
@@ -118,7 +118,10 @@ where
-        stack.push_stackframe(func_idx, locals, 0);
+
+        // setting `usize::MAX` as return address for the outermost function ensures that we
+        // observably fail upon errornoeusly contining execution after that function returns.
```
typo: `contining`

wucke13 commented:
With the git wrangling I rewrote all commits that @george-cosma contributed to. I added them as co-author however, to correctly reflect that these commits contain significant contribution from them.

I hope, that this is now finally ready for merge.

valexandru commented on src/execution/interpreter_loop.rs:
```diff
@@ -68,26 +60,68 @@ pub(super) fn run<H: HookSet>(
-                break;
+                let maybe_return_address = stack.pop_stackframe();
+
+                // We finished this entire invocation if there is no tackframe left. If there are
```
```suggestion
                // We finished this entire invocation if there is no stackframe left. If there are
```

wucke13 replied:
Good catch, fixed!

valexandru commented on src/execution/value_stack.rs:
```diff
@@ -12,81 +13,174 @@ use crate::unreachable_validated;
+        // stackframe. However, there is one legitimate reason to pop when there is **no** current
+        // stackframe: after the outermost function returns, to extract the final return values of
+        // this interpreter invocation.
+        debug_assert!(
```
A question here, maybe we discussed about this, but would we want an error state/panic/etc. in the release mode if we manage to get here?

wucke13 replied:
Yes this is only there to find bugs when running `cargo test`. Basically, assuming that there are no logic bugs in the implementation, this will never panic.

wucke13 replied:
It did show me an intersting logic bug though: I inserted a `debug_assert` to verify that values are never popped from the value stack past the current stack frames base pointer. Turns out, there is a moment where there is 0 `CallFrame`s, but is still legal to pop values from the stack: after an invocation finished, to retrieve the return values.

cemonem commented:
implemented in https://github.com/DLR-FT/wasm-interpreter/blob/400533d972c06ab8d16afee759417cb40c7ca5c7/src/execution/interpreter_loop.rs#L114
